### PR TITLE
Support GDI+

### LINKED ISSUE #207: Add bindings for native GDI+
_Most_ of GDI+ is already available in System.Drawing, but there are a few things in GDI+ 1.1 that aren't. Namely, the effects API in `gdipluseffects.h`. I use the red-eye correction effect (`Gdiplus::RedEyeCorrection`) in Paint.NET, and it'd be nice to move that block of code out of C and into C#.

This is a nice-to-have and not a high priority, especially since it requires handling a bunch of C++ header files.

### PR DESCRIPTION
**Pull request by reflectronic**
Closes #207 

Does not attempt to add bindings for C++ helpers. Many manual changes were involved in generating these headers; some ClangSharp fixes will be in order.

### CODE REVIEW

File: sources/Interop/Windows/um/gdipluscolor/Color.cs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            return ((uint)((uint)(unchecked((byte)(GetRed())) | unchecked((ushort)((byte)(GetGreen())) << 8)) | (((uint)((byte)(GetBlue()))) << 16)));
+        }
+
+        [return: NativeTypeName("Gdiplus::ARGB")]
+        public static uint MakeARGB([NativeTypeName("BYTE")] byte a, [NativeTypeName("BYTE")] byte r, [NativeTypeName("BYTE")] byte g, [NativeTypeName("BYTE")] byte b)
+        {
+            return (((uint)(b) << (int)(BlueShift)) | ((uint)(g) << (int)(GreenShift)) | ((uint)(r) << (int)(RedShift)) | ((uint)(a) << (int)(AlphaShift)));
+        }
+
+        public const int AliceBlue = unchecked((int)(0xFFF0F8FF));
```
[the remark this answers is not in the document]
Comment (AraHaan) [reply sources/Interop/Windows/um/gdipluscolor/Color.cs:104]: Is there plans to include RebeccaPurple in with the other gdi+ colors?

**Comment by rickbrew** [comment]
I've been porting all of the native code in Paint.NET over to C#, and I'm almost done. I've got one small use of GDI+ that this PR would cover and then I could just delete my `PaintDotNet.SystemLayer.Native.{arch}.dll`. That would be really nice :)

Now that TerraFX has namespaces for `Windows`, `DirectX`, and `WinRT`, I think putting GDI+ into `TerraFX.Interop.Gdiplus` is a natural fit.

**Comment by AraHaan** [comment]
I think putting it in ``TerraFX.Interop.Windows.Gdiplus`` for it would be more fitting.

**Comment by tannergooding** [comment]
Closing in favor of https://github.com/terrafx/terrafx.interop.windows/pull/298

File: sources/Interop/Windows/um/gdiplusflat/Windows.cs [hunk 2]
```diff
@@ -0,0 +1,2504 @@
...
+using System;
+using System.Runtime.InteropServices;
+
+namespace TerraFX.Interop
+{
+    public static unsafe partial class Windows
+    {
+        [DllImport("GdiPlus", ExactSpelling = true)]
+        [return: NativeTypeName("Gdiplus::GpStatus")]
+        public static extern Status GdipCreatePath([NativeTypeName("Gdiplus::GpFillMode")] FillMode brushMode, [NativeTypeName("Gdiplus::GpPath **")] IntPtr* path);
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply sources/Interop/Windows/um/gdiplusflat/Windows.cs:15]: These are strongly typed in #298

File: sources/Interop/Windows/um/gdiplustypes/RectF.Manual.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+using System.Runtime.CompilerServices;
+using static TerraFX.Interop.Windows;
+
+namespace TerraFX.Interop
+{
+    public partial struct RectF
+    {
+        [return: NativeTypeName("Gdiplus::RectF *")]
+        public unsafe RectF* Clone()
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply sources/Interop/Windows/um/gdiplustypes/RectF.Manual.cs:14]: This is part of the GDI+ API but temporarily excluded from #298 until the allocation/freeing and general C++ support can be figured out.

File: sources/Interop/Windows/um/gdipluseffects/Windows.cs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+                };
+
+                Debug.Assert(data.Length == Unsafe.SizeOf<Guid>());
+                return ref Unsafe.As<byte, Guid>(ref MemoryMarshal.GetReference(data));
+            }
+        }
+
+        [DllImport("GdiPlus", ExactSpelling = true)]
+        [return: NativeTypeName("Gdiplus::Status")]
+        public static extern Status GdipCreateEffect([NativeTypeName("const GUID")] Guid guid, [NativeTypeName("Gdiplus::CGpEffect **")] IntPtr* effect);
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply sources/Interop/Windows/um/gdipluseffects/Windows.cs:292]: These are strongly typed in #298

File: sources/Interop/Windows/um/gdipluscolor/Color.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tannergooding) [reply sources/Interop/Windows/um/gdipluscolor/Color.cs:104]: These are `uint` in #298 

As for `RebeccaPurple`, this only mirrors the GDI+ headers. It doesn't aim to extend or improve it.

File: sources/Interop/Windows/um/gdiplustypes/Rect.cs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        }
+
+        [return: NativeTypeName("BOOL")]
+        public int IsEmptyArea()
+        {
+            return ((Width <= 0) || (Height <= 0)) ? 1 : 0;
+        }
+
+        [return: NativeTypeName("BOOL")]
+        public unsafe int Equals([NativeTypeName("const Gdiplus::Rect &")] Rect* rect)
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply sources/Interop/Windows/um/gdiplustypes/Rect.cs:89]: Friendly overloads are "out of scope" for the time being. They might be added eventually, but its a significant effort to make happen and covers far more than just GDI+

**Pull request closed without merging**